rclcpp/src/rclcpp/generic/typesupport_helpers.cpp

```diff
@@ -0,0 +1,155 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rclcpp/generic/typesupport_helpers.hpp"
+
+#include <memory>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <tuple>
+#include <utility>
+
+#include "ament_index_cpp/get_resources.hpp"
+#include "ament_index_cpp/get_package_prefix.hpp"
+
+#include "rcpputils/shared_library.hpp"
+
+#include "rosidl_runtime_cpp/message_type_support_decl.hpp"
+
+namespace rclcpp_generic
+{
+
+std::string get_typesupport_library_path(
+  const std::string & package_name, const std::string & typesupport_identifier)
+{
+  const char * filename_prefix;
+  const char * filename_extension;
+  const char * dynamic_library_folder;
+#ifdef _WIN32
+  filename_prefix = "";
+  filename_extension = ".dll";
+  dynamic_library_folder = "/bin/";
+#elif __APPLE__
+  filename_prefix = "lib";
+  filename_extension = ".dylib";
+  dynamic_library_folder = "/lib/";
+#else
+  filename_prefix = "lib";
+  filename_extension = ".so";
+  dynamic_library_folder = "/lib/";
+#endif
+
+  std::string package_prefix;
+  try {
+    package_prefix = ament_index_cpp::get_package_prefix(package_name);
+  } catch (ament_index_cpp::PackageNotFoundError & e) {
+    throw std::runtime_error(e.what());
+  }
+
+  auto library_path = package_prefix + dynamic_library_folder + filename_prefix +
+    package_name + "__" + typesupport_identifier + filename_extension;
+  return library_path;
+}
+
+const std::pair<std::string, std::string> extract_type_and_package(const std::string & full_type)
+{
+  std::string package_name;
+  std::string type_name;
+
+  std::tie(package_name, std::ignore, type_name) = extract_type_identifier(full_type);
+
+  return {package_name, type_name};
+}
+
+std::tuple<std::string, std::string, std::string>
+extract_type_identifier(const std::string & full_type)
+{
+  char type_separator = '/';
+  auto sep_position_back = full_type.find_last_of(type_separator);
+  auto sep_position_front = full_type.find_first_of(type_separator);
+  if (sep_position_back == std::string::npos ||
+    sep_position_back == 0 ||
+    sep_position_back == full_type.length() - 1)
+  {
+    throw std::runtime_error(
+            "Message type is not of the form package/type and cannot be processed");
+  }
+
+  std::string package_name = full_type.substr(0, sep_position_front);
+  std::string middle_module = "";
+  if (sep_position_back - sep_position_front > 0) {
+    middle_module =
+      full_type.substr(sep_position_front + 1, sep_position_back - sep_position_front - 1);
+  }
+  std::string type_name = full_type.substr(sep_position_back + 1);
+
+  return std::make_tuple(package_name, middle_module, type_name);
+}
+
+std::shared_ptr<rcpputils::SharedLibrary>
+get_typesupport_library(const std::string & type, const std::string & typesupport_identifier)
```

Comment:
In contrast, I think there is value in this function existing at the rclcpp level because it can make use of `rcpputils`'s version of shared library loading, which is safer to use than the C version provided by `rcutils`. But we still might want a version of this in `rcl` so other languages (Python, Java, Rust, etc.) could reuse it.